File: src/primes.ml
let max = int_of_string Sys.argv.(1)

let rec for_step f i j s =
  if i >= j then ()
  else (f i; for_step f (i+s) j s)

let _ =
  let nums = Array.make max false in

  (* 0 and 1 are not prime *)
  Array.set nums 0 true;
  Array.set nums 1 true;

  (* Computing prime numbers *)
  for i = 2 to int_of_float (sqrt (float_of_int max)) do
    if Array.get nums i = false then
      for_step (fun i -> Array.set nums i true) (i*i) max i
  done;

  (* Counting the prime numbers *)
  let total = Array.fold_left (fun acc n -> acc + (if n then 0 else 1)) 0 nums in

  Printf.printf "%d" total

File: src/primes.py
#!/usr/bin/python3

import sys # argv
import math # sqrt

MAX = int(sys.argv[1])

nums = [False] * MAX

# 0 and 1 are not prime
nums[0] = nums[1] = True

# Computing prime numbers
for i in range(2,int(math.sqrt(MAX))+1):
    if nums[i] == True:
        continue
    for j in range(i*i,MAX,i):
        nums[j] = True;

# Counting prime numbers
total = 0
for i in nums:
    total += i == False
print(total)
